File: Python/113.Path Sum II .py
#求出所有从根节点到叶节点的路径，路径上的数之和为目标数
import copy
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        self.sum = sum
        self.result , tmp = [] , []
        self.inorder(root,0,tmp)
        return self.result

    def inorder(self,p,cursum,tmp):
        if p!=None:
            cursum += p.val
            tmp.append(p.val)
            if p.left==None and p.right==None:
                if cursum==self.sum:
                    self.result.append(copy.deepcopy(tmp))
            else:
                self.inorder(p.left,cursum,tmp)
                self.inorder(p.right,cursum,tmp)
            tmp.pop()



t1 = TreeNode(5)
t2 = TreeNode(4)
t3 = TreeNode(8)
t4 = TreeNode(11)
t5 = TreeNode(13)
t6 = TreeNode(4)
t7 = TreeNode(7)
t8 = TreeNode(2)
t9 = TreeNode(5)
t1.left = t2
t1.right = t3
t2.left = t4
t3.right = t6
t3.left = t5
t4.left = t7
t4.right = t8
t6.right = t9

s = Solution()
print(s.hasPathSum(t1,22))

File: C++/Tree/113.Path Sum II.cpp
/*
找出从根到叶的所有路径，结点和等于给出的数。
*/
/*
思路：向下递归，把当前结点的值减掉传入下一层，记录中间路径结果，
	  到达叶结点时判断是否等于传入的值即可。
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
	vector<vector<int>> result;
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        if(!root)	return result;
        vector<int> tmp;
        helpPathSum(root,sum,tmp);
        return result;
    }
	void helpPathSum(TreeNode* root, int sum,vector<int>& tmp)
	{
		//if(root->val > sum)	return;  //不能提前终止，因为结点值有负的。
		tmp.push_back(root->val);
		if(!root->left && !root->right && root->val==sum)
			result.push_back(tmp);
		if(root->left)	helpPathSum(root->left,sum-root->val,tmp);
		if(root->right) helpPathSum(root->right,sum-root->val,tmp);
		tmp.pop_back();
	}
};
int main()
{
	Solution s;
	TreeNode root(1),t1(2),t2(3);
	root.right = &t1;
	t1.left = &t2;
	vector<vector<int>> out = s.pathSum(&root,6);
}

